reconcile/external_resources/state.py

```diff
@@ -0,0 +1,247 @@
+            resource_status=self._get_value(item, self.RESOURCE_STATUS),
+            reconciliation=r,
+            reconciliation_errors=int(
+                self._get_value(item, self.RECONCILIATION_ERRORS, _type="N")
+            ),
+        )
+
+    def serialize(self, state: ExternalResourceState) -> dict[str, Any]:
+        return {
+            self.ER_KEY_HASH: {"S": state.key.hash()},
+            self.TIMESTAMP: {"S": state.ts.isoformat()},
+            self.RESOURCE_STATUS: {"S": state.resource_status.value},
+            self.RECONCILIATION_ERRORS: {"N": str(state.reconciliation_errors)},
+            self.ER_KEY: {
+                "M": {
+                    self.ER_KEY_PROVISION_PROVIDER: {"S": state.key.provision_provider},
+                    self.ER_KEY_PROVISIONER_NAME: {"S": state.key.provisioner_name},
+                    self.ER_KEY_PROVIDER: {"S": state.key.provider},
+                    self.ER_KEY_IDENTIFIER: {"S": state.key.identifier},
+                }
+            },
+            self.RECONC: {
+                "M": {
+                    self.RECONC_RESOURCE_HASH: {
+                        "S": state.reconciliation.resource_hash
+                    },
+                    self.RECONC_ACTION: {"S": state.reconciliation.action.value},
+                    self.RECONC_INPUT: {"S": state.reconciliation.input},
+                    self.MODCONF: {
+                        "M": {
+                            self.MODCONF_IMAGE: {
+                                "S": state.reconciliation.module_configuration.image
+                            },
+                            self.MODCONF_VERSION: {
+                                "S": state.reconciliation.module_configuration.version
+                            },
+                            self.MODCONF_DRIFT_MINS: {
+                                "N": str(
+                                    state.reconciliation.module_configuration.reconcile_drift_interval_minutes
+                                )
+                            },
+                            self.MODCONF_TIMEOUT_MINS: {
+                                "N": str(
+                                    state.reconciliation.module_configuration.reconcile_timeout_minutes
+                                )
+                            },
+                        }
+                    },
+                }
+            },
+        }
+
+
+class ExternalResourcesStateDynamoDB:
+    PARTIALS_PROJECTED_VALUES = ",".join([
+        DynamoDBStateAdapter.ER_KEY,
+        DynamoDBStateAdapter.TIMESTAMP,
+        DynamoDBStateAdapter.RESOURCE_STATUS,
+        DynamoDBStateAdapter.RECONCILIATION_ERRORS,
+        f"{DynamoDBStateAdapter.RECONC}.{DynamoDBStateAdapter.RECONC_RESOURCE_HASH}",
+    ])
+
+    def __init__(self, table_name: str, index_name: str, region_name: str) -> None:
+        self.adapter = DynamoDBStateAdapter()
+        self.client = boto3.client("dynamodb", region_name=region_name)
```

Comment:
not sure if this client will create a new session that need to be closed after use, can add the cleanup later once we find it's leaking.